### PR TITLE
[KYUUBI #1798] Add EventBus module to unify the distribution and subscription of Kyuubi's events

### PR DESCRIPTION
**Pull request by iodone**
### _Why are the changes needed?_

Use case is simple:
```
    // scribe sync
    EventBus.register[Test0KyuubiEvent](new Test0Handler with EventHandler)
   
    // scribe async
     EventBus.registerAsync[Test0KyuubiEvent](new Test0Handler with EventHandler)

    // send event
    EventBus.post(KyuubiEvent("test0"))
    EventBus.post(KyuubiEvent("test1"))
```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-download/pom.xml [hunk 1]
```diff
@@ -45,6 +45,7 @@
                         </goals>
                         <configuration>
                             <skip>${spark.archive.download.skip}</skip>
+                            <skipCache>true</skipCache>
```
Comment (yaooqinn) [inline externals/kyuubi-download/pom.xml:48]: this seems irrelevant

**Comment by yaooqinn** [comment]
cc @wForget, can you help check this PR?

File: kyuubi-server/src/main/resources/kyuubi-defaults.conf [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#
```
Comment (yaooqinn) [inline kyuubi-server/src/main/resources/kyuubi-defaults.conf:1]: this file is unnecessary?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  def registerAsync[T <: KyuubiEvent: ClassTag](et: EventHandler[T]): EventBus =
+    defaultEventBus.registerAsync[T](et)
+
+  private case class EventBusLive() extends EventBus {
+    private[this] val eventHandlerRegistry = new Registry
+    private[this] val asyncEventHandlerRegistry = new Registry
+    implicit private[this] val ec: ExecutionContext =
+      ExecutionContext.fromExecutor(
+        ThreadUtils.newDaemonQueuedThreadPool(10, 10, 60000L, "event-bus-exec-pool"))
```
Comment (wForget) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala:55]: The queue size seems to be too small, and how do we handle subsequent events when the queue is full?

**Comment by wForget** [comment]
> cc @wForget, can you help check this PR?

Sorry, I'm confused about this, this is overcorrect for my needs.
For me, I just need to store the event asynchronously and don't need to add some handlers to handle it when producing events. I will read the event in storage(like es), then process and audit it.

In addition, I feel this is similar to before:
1. EventBus is similar to EventLoggingService
2. EventHandler is similar to EventLogger

**Comment by iodone** [comment]
> > cc @wForget, can you help check this PR?
> 
> Sorry, I'm confused about this, this is overcorrect for my needs. For me, I just need to store the event asynchronously and don't need to add some handlers to handle it when producing events. I will read the event in storage(like es), then process and audit it.
> 
> In addition, I feel this is similar to before:
> 
> 1. EventBus is similar to EventLoggingService
> 2. EventHandler is similar to EventLogger

@wForget The motivation for the implementation of eventBus can be found in #1798 

EventBus is just a Handler for triggering and adding events, it does not contain any business logic, EventLoggerService can be seen as a concrete business implementation of one of these Handlers. Because the previous EventLoggerService implementation was slightly complex and coupled, it was adapted using EventBus in order to be compatible with the previous implementation logic, and at the same time, it is very clear and simple to use EventBus.

File: externals/kyuubi-download/pom.xml (hunk 1 above)
Comment (iodone) [reply externals/kyuubi-download/pom.xml:48]: A small problem found during testing, need to submit a new issue to fix?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala (hunk 3 above)
Comment (iodone) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala:55]: For asynchronous processing is relatively simple to implement, the size of the thread pool is configured by kyuubi Config or changed to a ForkJoin thread pool?

**Comment by yaooqinn** [comment]
Now, we have EventHandler/EventLogger/EventBus/EventBusLive/EventLoggingService, it's confusing. Maybe we can make the new design to keep it simple.

**Comment by yaooqinn** [comment]
The `EventLogger` s, such as `JsonEventLogger` extends `AbstractService` now which makes it hard to developers to implement.

And the EventLoggingService is now sync and we shall make it async simply.

WDYT? @wForget @iodone

**Comment by wForget** [comment]
> The `EventLogger` s, such as `JsonEventLogger` extends `AbstractService` now which makes it hard to developers to implement. we shall make it AbstractService-free.
> 
> And the EventLoggingService is now sync and we shall make it async simply.
> 
> WDYT? @wForget @iodone

+1, I prefer this.

**Comment by iodone** [comment]
> The `EventLogger` s, such as `JsonEventLogger` extends `AbstractService` now which makes it hard to developers to implement. we shall make it AbstractService-free.
> 
> And the EventLoggingService is now sync and we shall make it async simply.
> 
> WDYT? @wForget @iodone

This `AbstractService` is actually the `EventHandler` implemention in `EventBus`

**Comment by iodone** [comment]
image
The above is the overall design of the `EventBus` implementation, there is no relationship with the `EventLoggerService`, `EventLoggerService` is seen as a specific implementation of the `EventHandler`. `EventBus` main functions: 
1. Register a `KyuubiEvent` eventHandler, where the `EventHandler` can be `EventLoggerService`, `AuditService`, or any other logic for handling events
2. Trigger an event at some place

By adding `EventBus` is mainly to decouple `EventLoggerService`, I think `EventLoggerService` is only suitable to handle events, it is just an `EventHandler`, it should not trigger event, this is the responsibility of `EventBus`.

**Comment by iodone** [comment]
The definition of `EventHandler` is a function with few constraints, and the input is `KyuubiEvent`, which is a fully AbstractService-free:
```scala
package object events {
  type EventHandler[T <: KyuubiEvent] = T => Unit
}
```
We can implement `EventHandler`, either following the `AbstractService` pattern, or freely writing any implementation pattern.
In this commit it is easy to convert the previously implemented `EventLoggerService` into an `EventHandler` without any intrusiveness.
```scala
 private val eventLoggingService =
    new EventLoggingService with EventHandler[KyuubiServerEvent] {
      override def apply(event: KyuubiServerEvent): Unit = onEvent(event)
    }
```
@yaooqinn @wForget Does this explanation for `EventBus` meet your expectations?

**Comment by yaooqinn** [comment]
Sorry for the late reply.

the idea is great, but can we make it a bit simple. As `EventLogger` can be a typical `EventHandler`, we can remove one layer(EventLoggerService) directly.

**Comment by iodone** [comment]
> Sorry for the late reply.
> 
> the idea is great, but can we make it a bit simple. As `EventLogger` can be a typical `EventHandler`, we can remove one layer(EventLoggerService) directly.

OK，i will improve it

**Comment by iodone** [comment]
@yaooqinn , Could you help me review the code?

**Comment by yaooqinn** [comment]
@iodone thanks for the update. I will review this ASAP. also cc @wForget @ulysses-you

**Comment by wForget** [comment]
LGTM, in addition, I have some comments:
1. Why don't we register handlers with `EventBus.registerAsync`?
2. `SparkJsonLoggingEventHandler` and `ServerJsonLoggingEventHandler` don't seem to do anything, they seem unnecessary.
3. Do we need to use a separate `ExecutionContext` for EventBus instead of `Implicits.global`?

**Comment by iodone** [comment]
> LGTM, in addition, I have some comments:
> 
> 1. Why don't we register handlers with `EventBus.registerAsync`?
> 2. `SparkJsonLoggingEventHandler` and `ServerJsonLoggingEventHandler` don't seem to do anything, they seem unnecessary.
> 3. Do we need to use a separate `ExecutionContext` for EventBus instead of `Implicits.global`?
 
@wForget, thank you for your advice. 
For points 1, 3:
Currently, since the Async pattern has not been validated in real scenarios: 
1. Just the same implementation of the previous LoggerService is done, not upgraded to asynchronous.
2. Scala global `executioncontext` uses the `ForkJoin` settings related to the system, which can ensure optimal performance. We can tuning it when using async EventHandler in the future like the implement of `AsyncElasticSearchLoggingHandler`

For points 2:
For decoupling and extensibility, while also facilitating independent unit testing

**Comment by wForget** [comment]
> > LGTM, in addition, I have some comments:
> > 
> > 1. Why don't we register handlers with `EventBus.registerAsync`?
> > 2. `SparkJsonLoggingEventHandler` and `ServerJsonLoggingEventHandler` don't seem to do anything, they seem unnecessary.
> > 3. Do we need to use a separate `ExecutionContext` for EventBus instead of `Implicits.global`?
> 
> @wForget, thank you for your advice. For points 1, 3: Currently, since the Async pattern has not been validated in real scenarios:
> 
> 1. Just the same implementation of the previous LoggerService is done, not upgraded to asynchronous.
> 2. Scala global `executioncontext` uses the `ForkJoin` settings related to the system, which can ensure optimal performance. We can tuning it when using async EventHandler in the future like the implement of `AsyncElasticSearchLoggingHandler`
> 
> For points 2: For decoupling and extensibility, while also facilitating independent unit testing

yea, these are some small doubts. The overall implementation looks good. @yaooqinn @ulysses-you  WDYT?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 4]
```diff
@@ -170,15 +169,37 @@ object SparkSQLEngine extends Logging {
...
+
+    def initLoggerEventHandler(conf: KyuubiConf): Unit = {
+      conf.get(ENGINE_EVENT_LOGGERS)
+        .map(EventLoggerType.withName)
+        .foreach {
+          case EventLoggerType.SPARK =>
+            EventBus.register[KyuubiEvent](new SparkHistoryLoggingEventHandler(spark.sparkContext))
+          case EventLoggerType.JSON =>
+            val handler = SparkJsonLoggingEventHandler(
+              spark.sparkContext.applicationAttemptId.getOrElse(spark.sparkContext.applicationId),
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:189]: `applicationAttemptId` is a number of attempt count, e.g. `1`

The right name maybe:
```
spark.sparkContext.applicationAttemptId
  .map(id => s"${spark.sparkContext.applicationId}_$id")
  .getOrElse(spark.sparkContext.applicationId)
```

Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:189]: I see this code is coied, just fixes it together

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 5]
```diff
@@ -170,15 +169,37 @@ object SparkSQLEngine extends Logging {
...
+              spark.sparkContext.hadoopConfiguration,
+              conf)
+
+            // register JsonLogger as a event handler for default event bus
+            EventBus.register[KyuubiEvent](handler)
+          case logger =>
+            // TODO: Add more implementations
+            throw new IllegalArgumentException(s"Unrecognized event logger: $logger")
+        }
+
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:200]: unnecessary line

File: kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.reflect.{classTag, ClassTag}
+import scala.util.Try
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.events.handler.EventHandler
+
+sealed trait EventBus {
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala:29]: can we enrich the class description ?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  def post[T <: KyuubiEvent](event: T): Unit
+
+  def register[T <: KyuubiEvent: ClassTag](et: EventHandler[T]): EventBus
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala:32]: why this and some other place name it `et`...  how about `eventHander` ?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+  def register[T <: KyuubiEvent: ClassTag](et: EventHandler[T]): EventBus =
+    defaultEventBus.register[T](et)
+
+  def registerAsync[T <: KyuubiEvent: ClassTag](et: EventHandler[T]): EventBus =
+    defaultEventBus.registerAsync[T](et)
+
+  private case class EventBusLive() extends EventBus {
+    private[this] val eventHandlerRegistry = new Registry
+    private[this] val asyncEventHandlerRegistry = new Registry
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/events/EventBus.scala:53]: add `lazy`

**Review by ulysses-you — COMMENTED** [review]
sorry for the late, overall looks good. thank you @iodone

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 4 above)
Comment (iodone) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:189]: Thanks, i will resolve it ASAP

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master
